src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs

```diff
@@ -4004,6 +4004,542 @@ public void M4(C? c)
+                //         _ = (b && M0(x = y = 0)) is bool // 1
+                Diagnostic(ErrorCode.WRN_IsAlwaysTrue, "(b && M0(x = y = 0)) is bool").WithArguments("bool").WithLocation(11, 13),
+                // (12,15): error CS0165: Use of unassigned local variable 'x'
+                //             ? x.ToString() // 2
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "x").WithArguments("x").WithLocation(12, 15),
+                // (13,15): error CS0165: Use of unassigned local variable 'y'
+                //             : y.ToString(); // 3
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "y").WithArguments("y").WithLocation(13, 15)
+                );
+        }
+
+        [Theory]
+        [InlineData("true", "false")]
+        [InlineData("!false", "!true")]
+        [InlineData("(true || false)", "(true && false)")]
+        public void IsBoolConstant_01(string @true, string @false)
+        {
+            var source = @"
+#nullable enable
+
+class C
+{
+    bool M0(object obj) => false;
+
+    void M1(bool b)
+    {
+        int x, y;
+        _ = (b && M0(x = y = 0)) is " + @true + @"
+            ? x.ToString()
+            : y.ToString(); // 1
+    }
+
+    void M2(bool b)
+    {
+        int x, y;
+        _ = (b && M0(x = y = 0)) is " + @false + @"
+            ? x.ToString() // 2
+            : y.ToString();
+    }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics(
+                // (13,15): error CS0165: Use of unassigned local variable 'y'
+                //             : y.ToString(); // 1
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "y").WithArguments("y").WithLocation(13, 15),
+                // (20,15): error CS0165: Use of unassigned local variable 'x'
+                //             ? x.ToString() // 2
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "x").WithArguments("x").WithLocation(20, 15)
+                );
+        }
+
+        [Fact]
+        public void IsBoolConstant_02()
+        {
+            var source = @"
+#nullable enable
+
+class C
+{
+    bool M0(object obj) => false;
+
+    void M1(bool b)
+    {
+        int x, y;
+        _ = (b && M0(x = y = 0)) is var z
+            ? x.ToString() // 1
+            : y.ToString(); // unreachable
+    }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics(
+                // (12,15): error CS0165: Use of unassigned local variable 'x'
+                //             ? x.ToString() // 1
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "x").WithArguments("x").WithLocation(12, 15)
+                );
+        }
+
+        [Fact]
+        public void IsBoolConstant_03()
+        {
+            var source = @"
+#nullable enable
+
+class C
+{
+    bool M0(object obj) => false;
+
+    void M1(bool b)
+    {
+        int x, y;
+        _ = (b && M0(x = y = 0)) is not true
+            ? x.ToString() // 1
+            : y.ToString(); // 2
```

Comment:
Yeah, I really don't like this. `not` needs to be symmetrical.

Reply:
Fixed